src/da14531/da14531_protocol.c

```diff
@@ -0,0 +1,421 @@
+        crc = crc_finalize(crc);
+
+        self->state = SERIAL_LINK_STATE_READING;
+        self->frame_len = 0;
+
+        ASSERT(crc == crc_frame);
+        if (crc == crc_frame) {
+            return (struct da14531_protocol_frame*)&self->frame[0];
+        }
+        util_log("da14531: ERROR, invalid crc, dropped frame");
+    } break;
+    default:
+        break;
+    }
+    return NULL;
+}
+
+static void _serial_link_format_byte(uint8_t data, uint8_t* buf, uint16_t buf_len, uint16_t* idx)
+{
+    ASSERT(*idx + 2 < buf_len);
+    (void)buf_len;
+    switch (data) {
+    case SL_SOF:
+    case SL_ESCAPE:
+        buf[(*idx)++] = SL_ESCAPE;
+        buf[(*idx)++] = data ^ SL_XOR;
+        break;
+    default:
+        buf[(*idx)++] = data;
+        break;
+    }
+}
+
+uint16_t da14531_protocol_format(
+    uint8_t* buf,
+    uint16_t buf_len,
+    enum da14531_protocol_packet_type type,
+    const uint8_t* payload,
+    uint16_t payload_len)
+{
+    uint16_t idx = 0;
+    crc_t crc = crc_init();
+
+    ASSERT(idx + 1 < buf_len);
+    buf[idx++] = SL_SOF;
+
+    crc = crc_update(crc, &type, 1);
+    _serial_link_format_byte(type, buf, buf_len, &idx);
+
+    uint8_t len = payload_len & 0xff;
+    crc = crc_update(crc, &len, 1);
+    _serial_link_format_byte(len, buf, buf_len, &idx);
+
+    len = (payload_len >> 8) & 0xff;
+    crc = crc_update(crc, &len, 1);
+    _serial_link_format_byte(len, buf, buf_len, &idx);
+
+    for (int i = 0; i < payload_len; i++) {
+        _serial_link_format_byte(payload[i], buf, buf_len, &idx);
+    }
+
+    crc = crc_update(crc, &payload[0], payload_len);
+    crc = crc_finalize(crc);
+
+    // crc_t is the "fastest" type that holds u16, so can be longer than 2 bytes
+    for (unsigned int i = 0; i < sizeof(uint16_t); i++) {
+        _serial_link_format_byte(crc & 0xff, buf, buf_len, &idx);
+        crc >>= 8;
+    }
+    ASSERT(idx + 1 < buf_len);
+    buf[idx++] = SL_SOF;
+    return idx;
+}
+
+struct da14531_protocol_frame* da14531_protocol_poll(
+    uint8_t* in_buf,
+    uint16_t* in_buf_len,
+    const uint8_t* hww_data,
+    struct ringbuffer* out_queue)
+{
+    if (hww_data) {
+        util_log("out: %s", util_dbg_hex(hww_data, 64));
+        uint8_t tmp[128];
+        int len = da14531_protocol_format(
+            &tmp[0], sizeof(tmp), DA14531_PROTOCOL_PACKET_TYPE_BLE_DATA, hww_data, 64);
+        ASSERT(len < (int)sizeof(tmp));
+        ASSERT(ringbuffer_num(out_queue) + len <= out_queue->size);
+        for (int i = 0; i < len; i++) {
+            ringbuffer_put(out_queue, tmp[i]);
+        }
+    }
+    struct da14531_protocol_frame* frame = NULL;
+    if (_protocol.loader.state != FIRMWARE_LOADER_STATE_DONE) {
+        _firmware_loader_poll(&_protocol.loader, in_buf, in_buf_len, out_queue);
+    } else {
+        frame = _serial_link_in_poll(&_protocol.serial_link, in_buf, in_buf_len);
+    }
+    return frame;
+}
+
+static bool _swd_reset_da14531(void)
```

Comment:
guard with `#if FACTORYSETUP == 1 || !defined(NDEBUG)` like at the callsite, otherwise the compiler complains that this is unused (e.g. in `make firmware`)